### PR TITLE
Resolve Issue #1013 None radio values don't work as expected

### LINKED ISSUE #1013: UKIS: None options are getting preselected.
### Expected behaviour
'None' options must not get pre-selected.

### Actual behaviour
All the none options are getting pre-selected

### Steps to reproduce the behaviour
Launch an UKIS survey.
Navigate to Question 9.9 
### Screenshot
screen shot 2017-03-06 at 18 23 13

### PR DESCRIPTION
**Pull request by collisdigital**
### What is the context of this PR?
Fixes #1013

WTForms defaults to use a coerce function on Radio (Selector) fields of `str()`; this will convert Python `NoneType` to `'None'` string which means WTForms cannot tell the difference between the user not choosing any option `data=None` and choosing an option that happens to have a value of `value='None'` in HTML (i.e. `data='None'`).

This PR changes the coerce function for radio fields to be aware of `NoneType` in Python to avoid converting it to the string `'None'` to avoid this clash. This fixes #1013 which is caused by a Survey that contains a Radio field with a `'None'` option.

I have updated the Selenium tests and test_radio.json to exercise a number of scenarios around this (choosing None, changing the answer, checking the summary screen etc.)

I also added a None option to test_checkbox.json - but it doesn't appear affected by the same issue (presumably because it processes lists `[]` not just a single value).

### How to review 
- Run Selenium tests (or rely on Travis pass)
- Run some manual testing with radio fields 
- Check issue #1013 is resolved.

### CODE REVIEW

File: app/forms/fields.py [hunk 1]
```diff
@@ -104,14 +104,30 @@ def get_select_multiple_field(answer, label, guidance, error_messages):
...
+    :param value: Any value that can be coerced using str() or None
+    :return: str(value) or None if value is None
+    '''
+    return str(value) if value is not None else None
+
+
 def get_select_field(answer, label, guidance, error_messages):
     validate_with = get_mandatory_validator(answer, error_messages)
 
+    # We use a custom coerce function to avoid a defect where Python NoneType
```
Comment (benjefferies) [inline app/forms/fields.py:119]: Looks like this is a stab at fixing https://github.com/wtforms/wtforms/pull/288. Might be worth referencing it incase it gets merged and this can be undone

File: tests/app/forms/test_fields.py [hunk 2]
```diff
@@ -153,10 +153,20 @@ def test_radio_field(self):
...
         self.assertEquals(unbound_field.kwargs['choices'], expected_choices)
 
+    def test__coerce_str_unless_none(self):
+        # pylint: disable=protected-access
+        self.assertEquals(_coerce_str_unless_none(1), '1')
+        self.assertEquals(_coerce_str_unless_none('bob'), 'bob')
+        self.assertEquals(_coerce_str_unless_none(12323245), '12323245')
+        self.assertEquals(_coerce_str_unless_none('9887766'), '9887766')
+        self.assertEquals(_coerce_str_unless_none(None), None)
+        self.assertNotEquals(_coerce_str_unless_none(None), 'None')
```
Comment (benjefferies) [inline tests/app/forms/test_fields.py]: This assertion is redundant from the assertion above

File: tests/app/forms/test_fields.py [hunk 3]
```diff
@@ -153,10 +153,20 @@ def test_radio_field(self):
...
+        self.assertTrue(unbound_field.kwargs['coerce'], _coerce_str_unless_none)
         self.assertEquals(unbound_field.kwargs['label'], radio_json['label'])
         self.assertEquals(unbound_field.kwargs['description'], radio_json['guidance'])
         self.assertEquals(unbound_field.kwargs['choices'], expected_choices)
 
+    def test__coerce_str_unless_none(self):
+        # pylint: disable=protected-access
+        self.assertEquals(_coerce_str_unless_none(1), '1')
+        self.assertEquals(_coerce_str_unless_none('bob'), 'bob')
+        self.assertEquals(_coerce_str_unless_none(12323245), '12323245')
```
Comment (benjefferies) [inline tests/app/forms/test_fields.py:165]: Is there value in having this assertion as wel as `self.assertEquals(_coerce_str_unless_none(1), '1')`

File: tests/functional/generate_pages.py [hunk 4]
```diff
@@ -29,7 +29,7 @@
 
 HEADER = r"""// >>> WARNING THIS PAGE WAS AUTO-GENERATED - DO NOT EDIT!!! <<<
 
-import {basePage} from '../../{basePageFile}'
+import {basePage} from '../{basePageFile}'
```
Comment (benjefferies) [inline tests/functional/generate_pages.py:32]: Good effort updating this

File: tests/functional/generate_pages.py [hunk 5]
```diff
@@ -153,7 +160,7 @@ def process_answer(question_type, answer, page_spec):
         logger.debug("\t\tSkipping Child Answer: %s", answer['id'])
         return
     elif answer['type'] == 'Radio' or answer['type'] == 'Checkbox':
-        process_options(answer['id'], answer['options'], CHECKBOX_RADIO_CLICKER, page_spec)
+        process_options(answer['id'], answer['options'], CHECKBOX_RADIO_CLICKER, CHECKBOX_RADIO_IS_SELECTED, page_spec)
```
Comment (benjefferies) [inline tests/functional/generate_pages.py:165]: generate_camel_case_from_id(answer_id) ?

File: tests/functional/spec/radio.spec.js [hunk 6]
```diff
@@ -53,22 +53,94 @@ describe('Radio button with "other" option', function() {
...
     RadioMandatoryPage.clickRadioMandatoryAnswerBacon().submit()
-    RadioNonMandatory.clickTopPrevious()
+    RadioNonMandatoryPage.clickTopPrevious()
 
     //Then
     RadioMandatoryPage.clickOther()
     expect(RadioMandatoryPage.getOtherInputField()).to.equal('')
   })
 
+  it('Given I have previously selected the None option and saved, When I return to the screen, Then my choice is remembered.', function() {
```
Comment (benjefferies) [inline tests/functional/spec/radio.spec.js:74]: I wonder if it's easy enough to do this as a unit test with wtforms

File: tests/functional/spec/radio.spec.js [hunk 7]
```diff
@@ -53,22 +53,94 @@ describe('Radio button with "other" option', function() {
...
+
+    // When I return to the screen
+    RadioNonMandatoryPage.clickTopPrevious()
+
+    // Then my choice is remembered (Eggs)
+    expect(RadioMandatoryPage.RadioMandatoryAnswerNoneIsSelected()).to.be.false
+    expect(RadioMandatoryPage.RadioMandatoryAnswerBaconIsSelected()).to.be.false
+    expect(RadioMandatoryPage.RadioMandatoryAnswerEggsIsSelected()).to.be.true
+    expect(RadioMandatoryPage.RadioMandatoryAnswerSausageIsSelected()).to.be.false
+    expect(RadioMandatoryPage.RadioMandatoryAnswerOtherIsSelected()).to.be.false
```
Comment (benjefferies) [inline tests/functional/spec/radio.spec.js:115]: Does all of the checks below this add anything?

**Review by benjefferies — CHANGES_REQUESTED** [review]

File: tests/functional/spec/radio.spec.js (hunk 7 above)
Comment (collisdigital) [reply tests/functional/spec/radio.spec.js:115]: Prior to this fix it wasn't possible to change your answer from any option to the None option; so this verifies that you can change to None and then change to something else; corner cases but were real issues

File: app/forms/fields.py (hunk 1 above)
Comment (collisdigital) [reply app/forms/fields.py:119]: Good idea

File: tests/app/forms/test_fields.py (hunk 3 above)
Comment (collisdigital) [reply tests/app/forms/test_fields.py:165]: you never know; do multi-digit fields get treated differently?

File: tests/functional/spec/radio.spec.js (hunk 7 above)
Comment (benjefferies) [reply tests/functional/spec/radio.spec.js:115]: Fair enough. Does it warrant being a selenium test when the bug is in wtforms?

File: tests/functional/generate_pages.py (hunk 5 above)
Comment (collisdigital) [reply tests/functional/generate_pages.py:165]: Looks like answer_id is used on line 139 directly as well but I'll make some improvements to this as there is some duplication

File: tests/functional/spec/radio.spec.js (hunk 6 above)
Comment (collisdigital) [reply tests/functional/spec/radio.spec.js:74]: Possibly; I didn't try to be fair

File: app/forms/fields.py (hunk 1 above)
Comment (collisdigital) [reply app/forms/fields.py:119]: Will do

**Comment by collisdigital** [comment]
@benjefferies update force pushed (trying to keep this ready to merge)

File: app/forms/fields.py (hunk 1 above)
Comment (collisdigital) [reply app/forms/fields.py:119]: Done

File: tests/app/forms/test_fields.py [hunk 8]
```diff
@@ -153,10 +153,21 @@ def test_radio_field(self):
...
+        self.assertEquals(_coerce_str_unless_none('9887766'), '9887766')
+        self.assertEquals(_coerce_str_unless_none('None'), 'None')
```
Comment (collisdigital) [inline tests/app/forms/test_fields.py:167]: @ajmaddaford added 'None' check

File: tests/functional/generate_pages.py (hunk 4 above)
Comment (collisdigital) [reply tests/functional/generate_pages.py:32]: Although we should probably sort out the generation of the SummaryPage now it's a block; I updated to stop the script exploding on them but it produces an empty Summary for now.

File: tests/functional/generate_pages.py (hunk 5 above)
Comment (collisdigital) [reply tests/functional/generate_pages.py:165]: Pulled answer_name up to top of this function and used here also set option_name once in the process_options() instead of working out the name multiple times.

File: tests/functional/spec/radio.spec.js (hunk 7 above)
Comment (collisdigital) [reply tests/functional/spec/radio.spec.js:115]: We could do with a set of common scenarios for all question types; such as checking default value on first page load; submitting and returning to screen to see value persisted, changing the value, removing an answer etc. At the moment coverage of these typical use cases is patchy. If WTForms does change this issue could get re-introduced so perhaps safer to test our usage of it until we have sufficient use cases covered generically?

File: tests/app/forms/test_fields.py [hunk 9]
```diff
@@ -153,10 +153,21 @@ def test_radio_field(self):
...
+        self.assertEquals(_coerce_str_unless_none(None), None)
+        self.assertNotEquals(_coerce_str_unless_none(None), 'None')
```
Comment (LJBabbage) [inline tests/app/forms/test_fields.py]: Do we really need both? implicitly if something equals one thing, it won't equal another. We really don't need to test a string isn't equal to None do we?

self.assertEquals(_coerce_str_unless_none(None), None)
self.assertNotEquals(_coerce_str_unless_none(None), 'None')

**Review by LJBabbage — CHANGES_REQUESTED** [review]

File: tests/app/forms/test_fields.py (hunk 9 above)
Comment (benjefferies) [reply tests/app/forms/test_fields.py]: +1. We're testing pythons assertion library more than our code here

**Review by LJBabbage — APPROVED** [review]

**Comment by collisdigital** [comment]
@benjefferies force pushed a WTForms form test

File: tests/app/forms/test_questionnaire_form.py [hunk 10]
```diff
@@ -106,6 +106,69 @@ def test_date_range_to_precedes_from_raises_question_error(self):
         self.assertEqual(form.data, expected_form_data)
         self.assertEqual(form.question_errors['reporting-period-question'], expected_message)
 
+    def test_form_radio_options_no_selection(self):
```
Comment (benjefferies) [inline tests/app/forms/test_questionnaire_form.py:109]: Nice, really easy to read what these tests are doing/testing

File: tests/functional/generate_pages.py (hunk 5 above)
Comment (benjefferies) [reply tests/functional/generate_pages.py:165]: Probably should regenerate page models now I guess?

**Review by benjefferies — CHANGES_REQUESTED** [review]

**Review by benjefferies — APPROVED** [review]

**Comment by georgschoelly** [comment]
A related pull request was just merged in wtforms, this might solve the issue: https://github.com/wtforms/wtforms/pull/288

**Comment by nealedj** [comment]
Ah excellent. We'll look at updating wtforms and removing our workaround.
